connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/RestServerTest.java

```diff
@@ -392,6 +395,98 @@ public void testDisableAdminEndpoint() throws IOException {
+        WorkerConfig workerConfig = new DistributedConfig(workerProps);
+    }
+
+    @Test(expected = ConfigException.class)
+    public void testMissedHeaderName() {
+        String headerConfig = "add :DENY";
+        Map<String, String> workerProps = baseWorkerProps();
+        workerProps.put(WorkerConfig.RESPONSE_HTTP_HEADERS_CONFIG, headerConfig);
+        WorkerConfig workerConfig = new DistributedConfig(workerProps);
+    }
+
+    @Test(expected = ConfigException.class)
+    public void testMissedHeaderValue() {
+        String headerConfig = "add X-Frame-Options";
+        Map<String, String> workerProps = baseWorkerProps();
+        workerProps.put(WorkerConfig.RESPONSE_HTTP_HEADERS_CONFIG, headerConfig);
+        WorkerConfig workerConfig = new DistributedConfig(workerProps);
+    }
+
+    @Test(expected = ConfigException.class)
+    public void testInvalidHeaderConfigAction() {
+        String headerConfig = "badaction X-XSS-Protection: 1; mode=block";
+        Map<String, String> workerProps = baseWorkerProps();
+        workerProps.put(WorkerConfig.RESPONSE_HTTP_HEADERS_CONFIG, headerConfig);
+        WorkerConfig workerConfig = new DistributedConfig(workerProps);
+    }
+
+    public void checkCustomizedHttpResponseHeaders(String headerConfig, Map<String, String> expectedHeaders)
+            throws IOException  {
+        Map<String, String> workerProps = baseWorkerProps();
+        workerProps.put("offset.storage.file.filename", "/tmp");
+        workerProps.put(WorkerConfig.RESPONSE_HTTP_HEADERS_CONFIG, headerConfig);
+        WorkerConfig workerConfig = new DistributedConfig(workerProps);
+
+        EasyMock.expect(herder.kafkaClusterId()).andReturn(KAFKA_CLUSTER_ID);
+        EasyMock.expect(herder.plugins()).andStubReturn(plugins);
+        EasyMock.expect(plugins.newPlugins(Collections.emptyList(),
+                workerConfig,
+                ConnectRestExtension.class)).andStubReturn(Collections.emptyList());
+
+        EasyMock.expect(herder.connectors()).andReturn(Arrays.asList("a", "b"));
+
+        PowerMock.replayAll();
+
+        server = new RestServer(workerConfig);
+        server.initializeServer();
+        server.initializeResources(herder);
+        HttpRequest request = new HttpGet("/connectors");
+        CloseableHttpClient httpClient = HttpClients.createMinimal();
+        HttpHost httpHost = new HttpHost(server.advertisedUrl().getHost(), server.advertisedUrl().getPort());
+        CloseableHttpResponse response = httpClient.execute(httpHost, request);
+        Assert.assertEquals(200, response.getStatusLine().getStatusCode());
+        if (!headerConfig.isEmpty()) {
+            expectedHeaders.forEach((k, v) ->
+                    Assert.assertEquals(response.getFirstHeader(k).getValue(), v));
+        } else {
+            Assert.assertNull(response.getFirstHeader("X-Frame-Options"));
+        }
+        response.close();
+        server.stop();
+    }
+
```

Comment:
These are great and very useful, but they still don't thoroughly test the `validateHttpResponseHeaderConfig` and `validateHeaderConfigAction` methods. If you make these methods static, package-protected in the `WorkerConfig` class (rather than in a nested class), then you can easily add more tests to `WorkerConfigTest` that thoroughly verify all of the logic in those methods.

For example, something like the following in `WorkerConfigTest`:
```
    protected static final List<String> VALID_UNQUOTED_HEADER_CONFIGS = Arrays.asList(
            // TODO: Add a lot more valid header configs
            "\"add Cache-Control: no-cache, no-store, must-revalidate\"",
            "add X-XSS-Protection: 1; mode=block",
            "add Strict-Transport-Security: max-age=31536000; includeSubDomains",
            "AdD   Strict-Transport-Security:    max-age=31536000;  includeSubDomains",
            "AdD \t Strict-Transport-Security : \n   max-age=31536000;  includeSubDomains",
            "add X-Content-Type-Options: nosniff"
    );

    protected static final List<String> VALID_QUOTE_REQUIRED_HEADER_CONFIGS = Arrays.asList(
            // TODO: Add a lot more valid header configs
            "add Cache-Control: no-cache, no-store, must-revalidate"
    );

    protected static final List<String> INVALID_UNQUOTED_HEADER_CONFIGS = Arrays.asList(
            // TODO: Add a lot more valid header configs
            "WRONG Cache-Control: no-cache",
            "add Cache-Control no-cache",
            "WRONG Cache-Control: no-cache, no-store, must-revalidate",
    );

    protected static final String WHITESPACE = " \t \n  ";

    @Test
    public void testSingleValidHeaderConfigs() {
        for (String config : VALID_UNQUOTED_HEADER_CONFIGS) {
            assertValidHeaderConfig(config);
        }
        for (String config : VALID_QUOTE_REQUIRED_HEADER_CONFIGS) {
            assertValidHeaderConfig("\"" + config + "\"");
        }
    }

    @Test
    public void testSingleValidHeaderConfigsWithWhitespace() {
        for (String config : VALID_UNQUOTED_HEADER_CONFIGS) {
            assertValidHeaderConfig(WHITESPACE + config + WHITESPACE);
        }
        for (String config : VALID_QUOTE_REQUIRED_HEADER_CONFIGS) {
            assertValidHeaderConfig(WHITESPACE + "\"" + WHITESPACE + config + WHITESPACE + "\"" + WHITESPACE);
        }
    }

    @Test
    public void testMultipleValidHeaderConfigsWithoutWhitespace() {
        assertValidHeaderConfig(String.join(", ", VALID_UNQUOTED_HEADER_CONFIGS));
        assertValidHeaderConfig(String.join(" , ", VALID_UNQUOTED_HEADER_CONFIGS));
    }

    @Test
    public void testHeaderConfigsThatRequireQuotes() {
        for (String config : VALID_QUOTE_REQUIRED_HEADER_CONFIGS) {
            assertInvalidHeaderConfig(config);
        }
    }

    @Test
    public void testInvalidHeaderConfigs() {
        for (String config : INVALID_UNQUOTED_HEADER_CONFIGS) {
            assertInvalidHeaderConfig(config);
        }
    }

    @Test
    public void testOneInvalidAndMultipleValidHeaderConfigs() {
        assertInvalidHeaderConfig(String.join(", ", VALID_UNQUOTED_HEADER_CONFIGS)
                                  + ", " + INVALID_UNQUOTED_HEADER_CONFIGS.get(0));
        assertInvalidHeaderConfig(INVALID_UNQUOTED_HEADER_CONFIGS.get(0) + ", "
                                  + String.join(", ", VALID_UNQUOTED_HEADER_CONFIGS));
    }

    protected void assertValidHeaderConfig(String config) {
        WorkerConfig.validateHttpResponseHeaderConfig(config);
        // any valid config should be valid per HeaderFilter
        configureHeaderFilter(config);
    }

    protected void assertInvalidHeaderConfig(String config) {
        assertThrows(ConfigException.class, () -> WorkerConfig.validateHttpResponseHeaderConfig(config));
        // any invalid config should be also be invalid per HeaderFilter
        assertThrows(ConfigException.class, () -> configureHeaderFilter(config));
    }

    protected void configureHeaderFilter(String headerConfig) {
        FilterHolder headerFilterHolder = new FilterHolder(HeaderFilter.class);
        headerFilterHolder.setInitParameter("headerConfig", headerConfig);
        try {
            try {
                headerFilterHolder.doStart();
                headerFilterHolder.initialize();
            } finally {
                headerFilterHolder.doStop();
            }
        } catch (Exception e) {
          // wrap in ConfigException to keep the test simple
          throw new ConfigException("HeaderFilter failure", e);
        }
    }
```

Reply:
It is good idea  to assert each checking using assertValidHeaderConfig and assertInvalidHeaderConfig. I think latest version will cover validation of header config. Please let me know.

Reply:
Looks good. I like the additional checking that you're doing here.